R/slice.R

```diff
@@ -209,13 +216,25 @@ vec_assign <- function(x, i, value, ..., x_arg = "", value_arg = "") {
 # `[<-` with these inputs. In other words, we don't expect subclasses to have
 # vctrs subassign behavior, but we do expect them to match a subset of base R
 # subassign behavior.
-vec_assign_fallback <- function(x, i, value) {
-  # Work around bug in base `[<-`
-  existing <- !is.na(i)
-  i <- vec_slice(i, existing)
-  if (vec_size(value) != 1L) {
-    # Only slice `value` if we aren't recycling it
-    value <- vec_slice(value, existing)
+vec_assign_fallback <- function(x, i, value, location) {
+  value_size <- vec_size(value)
+
+  if (location) {
+    # Work around bug in base `[<-`
+    existing <- !is.na(i)
+    i <- vec_slice(i, existing)
+    if (value_size != 1L) {
+      # Only slice `value` if we aren't recycling it
+      value <- vec_slice(value, existing)
+    }
+  } else {
+    # Convert logical condition `i` to integer location `i`,
+    # and slice `value` to match. Treats `NA` as `FALSE`.
+    i <- which(i)
+    if (value_size != 1L) {
+      # Only slice `value` if we aren't recycling it
+      value <- vec_slice(value, i)
+    }
```

Comment:
In the fallback we convert to positive integer locations and slice `value` (like in my bench mark example)